plugins/janus_streaming.c

```diff
@@ -3394,62 +3521,289 @@ void janus_streaming_incoming_rtp(janus_plugin_session *handle, int video, char
 void janus_streaming_incoming_rtcp(janus_plugin_session *handle, int video, char *buf, int len) {
 	if(handle == NULL || g_atomic_int_get(&handle->stopped) || g_atomic_int_get(&stopping) || !g_atomic_int_get(&initialized))
 		return;
-	/* We might interested in the available bandwidth that the user advertizes */
-	uint64_t bw = janus_rtcp_get_remb(buf, len);
-	if(bw > 0) {
-		JANUS_LOG(LOG_HUGE, "REMB for this PeerConnection: %"SCNu64"\n", bw);
-		/* TODO Use this somehow (e.g., notification towards application?) */
-	}
 	janus_streaming_session *session = (janus_streaming_session *)handle->plugin_handle;
 	if(!session || session->destroyed || session->stopping || !session->started || session->paused)
 		return;
 	janus_streaming_mountpoint *mp = (janus_streaming_mountpoint *)session->mountpoint;
 	if(mp->streaming_source != janus_streaming_source_rtp)
 		return;
 	janus_streaming_rtp_source *ss = (janus_streaming_rtp_source *)mp->source;
+	janus_rtcp_header *rtcp = (janus_rtcp_header *)buf;
+	if(rtcp->version != 2)
+			return;
 
-	if ((video == 0) && (ss->audio_rtcp_fd > -1)) {
+	if((video == 0) && (ss->audio_rtcp_fd > -1)) {
 		JANUS_LOG(LOG_VERB, "Got Audio RTCP SSRC %u\n",
 			janus_rtcp_get_sender_ssrc(buf, len));
 
 		janus_rtcp_fix_ssrc(NULL, buf, len, 1,
 			janus_rtcp_get_sender_ssrc(buf, len),
 			ss->audio_ssrc);
 
-		struct sockaddr_in dest_rtcp;
-		dest_rtcp.sin_family = AF_INET;
-		dest_rtcp.sin_port = htons(ss->audio_rtcp_port);
-		dest_rtcp.sin_addr.s_addr = ss->audio_mcast;
-		int sent = 0;
+		char *filtered = NULL;
+		int newlen = 0;
+		int total = len, length = 0, bytes = 0;
+		/* Iterate on the compound packets */
+		gboolean keep = FALSE;
+		while(rtcp) {
+			keep = FALSE;
+			length = ntohs(rtcp->length);
+			if(length == 0)
+				break;
+			bytes = length*4+4;
+			switch(rtcp->type) {
+				case RTCP_RR:
+						/* incoming RR
+						 * check if one has been sent in the last minimum RR interval.
+						 * if not, update it with the worst packet loss fraction from the
+						 * last RR interval and send it along
+						 */
```

Comment:
I'm not really comfortable with how this works. While somehow aggregating PLI, REMB or other stuff from different users is fine, I feel it's broken for RR, which contains info that is necessarily specific to the viewer connection it refers to. Just messing with the loss fraction this way seems very prone to issues. Besides, I think RR makes sense only when we want to address the performance of the sender-to-mountpoint delivery itself, that is, a way for the plugin to tell the source "I missed three packets" (and hence check if there are connectivity problems with the source), rather than saying "someone somewhere on the public internet missed three packets". As it is, it will break or hinder delivery from components that would trust this information and limit themselves when we might not want them to.